src/erlcloud_cognito.erl

```diff
@@ -0,0 +1,906 @@
+    request(Config, "UpdateIdentityProvider", Body2).
+
+%%------------------------------------------------------------------------------
+%% Internal Functions
+%%------------------------------------------------------------------------------
+request(Config, Request) ->
+    Result = erlcloud_retry:request(Config, Request, fun handle_result/1),
+    case erlcloud_aws:request_to_return(Result) of
+        {ok, {_, <<>>}}     -> {ok, #{}};
+        {ok, {_, RespBody}} -> {ok, jsx:decode(RespBody, [return_maps])};
+        {error, _} = Error  -> Error
+    end.
+
+request(Config0, OperationName, Request) ->
+    case erlcloud_aws:update_config(Config0) of
+        {ok, Config} ->
+            Body       = jsx:encode(Request),
+            Operation  = "AWSCognitoIdentityProviderService." ++ OperationName,
+            Headers    = get_headers(Config, Operation, Body),
+            AwsRequest = #aws_request{service         = 'cognito-idp',
+                                      uri             = get_url(Config),
+                                      method          = post,
+                                      request_headers = Headers,
+                                      request_body    = Body},
+            request(Config, AwsRequest);
+        {error, Reason} ->
+            {error, Reason}
+    end.
+
+request_no_resp(Config, OperationName, Request) ->
+    case request(Config, OperationName, Request) of
+        {ok, _} -> ok;
+        Error   -> Error
+    end.
+
+get_headers(#aws_config{cognito_host = Host} = Config, Operation, Body) ->
+    Headers = [{"host", Host},
+        {"x-amz-target", Operation},
+        {"content-type", "application/x-amz-json-1.1"}],
+    Region = erlcloud_aws:aws_region_from_host(Host),
+    erlcloud_aws:sign_v4_headers(Config, Headers, Body, Region, "cognito-idp").
+
+handle_result(#aws_request{response_type = ok} = Request) ->
+    Request;
+handle_result(#aws_request{response_type   = error,
+                           error_type      = aws,
+                           response_status = Status} = Request)
+  when Status >= 500 ->
+    Request#aws_request{should_retry = true};
+handle_result(#aws_request{response_type = error,
+                           error_type    = aws} = Request) ->
+    Request#aws_request{should_retry = false}.
+
+get_url(#aws_config{cognito_scheme = Scheme,
+                    cognito_host   = Host}) ->
+    Scheme ++ Host.
+
+list_all(Fun, Args, Config, Key, TokenAlias) ->
+    list_all(Fun, Args, Config, Key, TokenAlias, undefined, []).
+
+list_all(Fun, Args, Config, Key, TokenAlias, NextToken, Acc) ->
+    UpdArgs = Args ++ [?MAX_RESULTS, NextToken, Config],
+    case erlang:apply(Fun, UpdArgs) of
+        {ok, Map} ->
+            UpdAcc   = Acc ++ maps:get(Key, Map),
+            NewToken = maps:get(TokenAlias, Map, undefined),
+            case NewToken of
+                undefined ->
+                    {ok, #{Key => UpdAcc}};
```

Comment:
e.g. https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_AdminListDevices.html

is this really a good way of returning list_all? does not it mingle the return format completely off the non list_all version?

Reply:
`list_all` returns data in same format as non list_all, but skips `PaginationToken` as it's not needed anymore. So in `AdminListDevices` example it'd look like
```
{
   "Devices": [ 
      ...
   ]
}
```

we can return just `UpdAcc` here and wrap it with `#{Key => ListAllResult}` later if that's what you mean

Reply:
probably i'm missing something. let me check one more thing